serverless/internal/git/git.go

```diff
@@ -0,0 +1,253 @@
+// Copyright 2021 Google LLC. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// git contains libraries for using github repositories that make serverless
+// operations easy to follow.
+package git
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/go-git/go-billy/v5/memfs"
+	"github.com/go-git/go-billy/v5/util"
+	"github.com/go-git/go-git/v5"
+	"github.com/go-git/go-git/v5/config"
+	"github.com/go-git/go-git/v5/plumbing"
+	"github.com/go-git/go-git/v5/plumbing/object"
+	"github.com/go-git/go-git/v5/storage/memory"
+	"github.com/golang/glog"
+	"github.com/google/go-github/v37/github"
+	"golang.org/x/oauth2"
+)
+
+// RepoID identifies a github project, including owner &repo.
+type RepoID struct {
+	Owner    string
+	RepoName string
+}
+
+// String returns "<owner>/<repo>".
+func (r RepoID) String() string {
+	return fmt.Sprintf("%s/%s", r.Owner, r.RepoName)
+}
+
+// NewRepoID creates a new RepoID struct from an owner/repo fragment.
+func NewRepoID(or string) (RepoID, error) {
+	s := strings.Split(or, "/")
+	if l, want := len(s), 2; l != want {
+		return RepoID{}, fmt.Errorf("can't split owner/repo %q, found %d parts want %d", or, l, want)
+	}
+	return RepoID{
+		Owner:    s[0],
+		RepoName: s[1],
+	}, nil
+}
+
+// NewServerlessRepo creates a wrapper around a git repository which has a fork owned by
+// the user, and an upstream repository configured that PRs can be proposed against.
+func NewServerlessRepo(ctx context.Context, upstream, fork RepoID, ghUser, ghEmail, ghToken, clonePath string) (ForkedRepo, error) {
+	repo := ForkedRepo{
+		upstream: upstream,
+		fork:     fork,
+		user:     ghUser,
+		email:    ghEmail,
+	}
+	var err error
+	repo.ghCli, err = authWithGithub(ctx, ghToken)
+	if err != nil {
+		return repo, err
+	}
+
+	// git clone -o origin "https://${GIT_USERNAME}:${FEEDER_GITHUB_TOKEN}@github.com/${fork_repo}.git" "${temp}"
+	cloneOpts := &git.CloneOptions{
+		URL: fmt.Sprintf("https://%v:%v@github.com/%s.git", ghUser, ghToken, repo.fork),
+	}
+	if clonePath == "" {
+		repo.git, err = git.Clone(memory.NewStorage(), memfs.New(), cloneOpts)
+	} else {
+		repo.git, err = git.PlainClone(clonePath, false, cloneOpts)
```

Comment:
This is my bad, but there are `CloneContext` and `PlainCloneContext` methods, for symmetric with `FetchContext` below.
Feel free to add a `TODO` on me to fix if you'd like.

Reply:
It's easier to fix than to write a TODO :-)